### ISSUE TITLE
A better way to maintain connection of libretro-database and libretro-thumbnails

### ISSUE THREAD

**Issue by guoyunhe** [libretro-thumbnails/libretro-thumbnails#118]
## The problem

1. ROM name is the only identifier to relate a ROM to a thumbnail.
2. ROM names come from different sources (nointro, redump, mame). And they are changed frequently.
3. When a ROM is renamed and the thumbnail doesn't, the connection breaks.
4. Scripts can help. But it still relies on guessing names and extra manual efforts.
5. Manage symlinks for Disc 1, Disc 2, Disc 3... are also painful.
6. Technical knowledge (Git, GitHub) are required when contributing to thumbnails.

## My solution

1. Create a web app with a SQL database.
2. Import ROM database and thumbnails. Create unique and persist IDs in SQL database for each record.
3. Add initial relation by matching names.
4. When upstream ROM name changes, but ROM hash/serial didn't change, the relation won't be affected.
5. Contributors can directly upload thumbnail on website. We have a filter to show which games' boxarts are missing.
6. Thumbnails can be exported back to GitHub, by month/week. It is possible to make the process fully automated.
7. With HTTP API, RetroArch users can search image with serial or hash(crc,md5,sha1) as fallback.

Some screenshots:

![图片](https://user-images.githubusercontent.com/5836790/213864076-1f829c2d-5317-4123-893a-9037ff3bef2b.png)

![图片](https://user-images.githubusercontent.com/5836790/213864543-d83f3b86-7309-4b48-906a-e24f027b29aa.png)

![图片](https://user-images.githubusercontent.com/5836790/213864552-817924c1-bf00-4d25-b498-7871baf79388.png)

![图片](https://user-images.githubusercontent.com/5836790/213864658-8375bb6c-f08e-4e3d-9f7f-4f37d59d7d7d.png)

- Back-end: https://github.com/guoyunhe/rgi-api
- Front-end: https://github.com/guoyunhe/rgi-web
- Demo: https://retrogameindex.netlify.app/

**Warning: the site is still in alpha, and data might be lost when upgrading!!!**

**Comment by RobLoach** [comment]
This is awesome! Nicely done! Thanks so much for the help here.

**Comment by i30817** [comment]
The thumbnail names dont _all_ come from the database. For instance, i'm adding names from the scummvm scanner into the thumbnails of scummvm, because those names are superior to grubby's scummvm pack, which is dead anyway. Frankly, i'm thinking of deleting those names because it would make things easier, although i'm not sure there.

Since the scummvm core doesn't scan normally, you have to either create scummvm files manually, or use a tool that takes scummvm.ini names and creates them for you (what i do) so i though i might as well get the 'good names' when creating the playlist.

**Comment by OctopusButtons** [comment]
"When a ROM is renamed and the thumbnail doesn't, the connection breaks."

I think it's more complicated and more confusing than that:  I've had cases where the thumbnail only downloads if I match the game name to no-intro NOT the libretro thumbnail name.  

RetroArch game name:  X
Libretro thumbnail name: X
No-intro game name: Y

Thumbnail doesn't download until I rename game in RetroArch to Y even though it's different from the thumbnail repository name.  The particular example was "Rev 1" versus "Rev A", and I've seen other cases with a coding/format change causing the same problem.  It's not the case.  Also saw it recently with NBA Jam titles, "_" versus "/" in the data separators between Libretro thumbnails and no-intro.

**Comment by i30817** [comment]
The algorithm to choose the thumbnails changed recently in a a way I don't like because it's less predictable (it chops off part of the local -label- name and tries to find that part in the server if it doesn't find the longest. This is less predictable because thumbnails are found for several categories, screenshot, title etc and this process of "finding" is done for each so you might end up with mixed screenshots of different languages for example, and in the same way if your name doesn't exist, for a region \language you might end up with a different region\language showing.

But as for your complaint, the names that appear on the playlist, no matter how they appear on the playlist are what is being used. So unless the scheme above affects you, which should not be possible for what you said, the way the name ended up in your label made it incorrect, because the thumbnails ONLY follow one of the dumping groups names (typically no intro for cartridges, and redump for CDs, with exceptions for other things like mame or scummvm). 

This can happen because of several reasons. You used automatic scanner and it gave you the wrong name because it's outdated, or you used manual scanner and had a different dumping group\outdated name

In NBA Jam _ is correct because / is a forbidden character. The algorithm is supposed for replace / among others in the label name searched by _ and the thumbnail incthe server is supposed to be _ too. Forbidden characters exist because they are forbidden in at least one of the filesystems RA is used on.


In this situation I usually use the manual scanner (it has a handy rescan button) and a external tool to rename rooms. I use rhdndat-rn here: https://pypi.org/project/rhdndat/ because I did the project, but I has limitations (only works on unziped roms, doesn't work in chds, if you want to check rvz (dolphin) roms requires a external executable and if you want to use a custom undo-hardpatch scheme I use to rename hardpatched hacks with a revert patch, you need xdelta on the path too (this is not necessary if you Don't want to rename hardpatched roms).

**Comment by i30817** [comment]
Tl;Dr: I don't trust the automatic scanner database names because they're often outdated from the REAL .dat from the dumping groups.

So I use the manual scanner (with dats when necessary like in MAME) to get real names, then i use a external tool I did, again with the dats to get updated names for the rom files when necessary before hitting rescan in playlist management and the names are updated, hopefully with the right thumb in the server too.

**Comment by i30817** [comment]
The only disadvantage of using the manual scanner is it has some usability traps (it's easy to get no thumbnails if you don't do certain things in the default options like having the RetroArch platform\database name for each root dir you scan), and it's more complex (you have to scan each platform one for one install of a single root dir for all). And the explore tag doesn't work (it barely works anyway, lots of missing games).

In return you scan all extensions you want as games, not just what libretro database does or does not recognize, and can use custom emulator entry point files, like dosbox .conf files or m3u files for multiple cd games, and can rescan.

**Comment by i30817** [comment]
Oh and another disadvantage is that Amiga whdload thumbnails are supported with a hack. Instead of doing another database just for thumbnails, what was made was link a "fixed" version of the whdload set to the current Amiga thumbnails. So if you scan a whdload set you get "Amiga floppy names" typically disc 1 even if it makes no sense....

This is a disadvantage because in the manual scanner you then get no thumbnails because the names are completely different (unless you do something else https://pypi.org/project/libretrofuzz/ which is also complex and has some false negatives\positives).

**Comment by i30817** [comment]
Finally, and this is out of topic please go vote for the democratic party if you're in the USA in these elections, I'm extremely worried about the fact the republican party is attempting a coup and the destruction of all public institutions if it succeeds. This WILL make you poorer and might even kill you in the future. That is all.

**Comment by OctopusButtons** [comment]
I thought a github method would be better than a separate web server, because the structure and system for contributions is already there.  (I might be missing some benefits of your system though.) I wish there was an addendum database used by RetroArch that maps hash/checksums to game versions in a github repository so that people could contribute checksums for language translation hacks, which then maps to some uniform (non-hack) game name, which would then get correct default thumbnail.  Though I see then the project members looking at the Pull Request have no good practical way to verify that the hash matches the supposed game named by the contributor.  

I think having to rely on organizations/systems like No Intro is a problem.   For example, No-Intro doesn't have a category for translations, because their **mindset** is "ROM warez scene: Dumps of physical products".  Libretro should have a **librarian mindset** where we have catalog match-ups for all  useful things, which definitely includes translations and usability/accessibility hacks which would then get appropriate default thumbnails.

> the thumbnails ONLY follow one of the dumping groups names (typically no intro for cartridges, and redump for CDs, [...].

The problem I saw was:

1. Libretro thumbnail respository file name: (Rev **A**)
2. Game name in playlist *and* ROM file name: (Rev **A**)
3. No thumbnail was auto-downloaded for that, while everything else in playlist is fine.
4. I looked on No-Intro's site and saw they had (Rev **1**) 
5. I manually change game name in RA to (Rev 1), and the thumbnail downloaded right away.  Even though that did NOT match the libretro repository file name.

Now I wonder if maybe there was a time lag between the thumbnail github repository and the thumbnail **server getting updated**? 

> [possibly] You used automatic scanner and it gave you the wrong name because it's outdated, or you used manual scanner and had a different dumping group\outdated name

I thought that the type of scan doesn't matter if automatic thumbnail download option is On, and I scroll to a game in a playlist and re-name it manually.  If I set the game name to match the libretro thumbnail repository file name, shouldn't that have downloaded the thumbnail?  I was doing Online Updater while troubleshooting so I don't think anything was outdated.

That case I'm talking about I don't intend as an issue, only as a piece of the discussion about the OP: a better way to maintain connectino of database and thumbnails.

**Comment by i30817** [comment]
Thumbnails are always downloaded from the label names. Or from the string derived from the label name. 

This is the hack (that I agree with!) that the project used to decouple how to search the thumbnail from how the thumbnail is scanned, because manual scan thumbnails don't have checksum\extra information, only the filename.


It just happens that the filename and the system (that you assign with the manual scanner) is enough to get to the right server directory.

It also allows you to have thumbnails for hacks\public domain new that thumbnail contributers bother to upload and aren't in no-intro. The rest is keeping up to date with no-intro names and names changes, which is a eternal struggle (especially the PRs to chane names here as soon as no intro changes something), but manageable to me with rhdndat.

**Comment by i30817** [comment]
There are disadvantages to the manual scanner, namely you lose the explore tab, if it worked for your game lists since many games have incomplete metadata, it could also be simplified slightly without loss of functionality and remove user traps.

But I think the advantages offset the disadvantages a lot. Just the rescan functionality is a major improvement, and making romhacks _actually appear_ in playlists is indispensable.

RetroArch also shot itself in the foot because of conflicting requirements and the normal scanner isn't actually checksumming anything these days (because people complained about their terabytes collections taking hours to scan). So it can't use that info to match games to database entries, can't avoid factory misprints of serials, identifies all hacks as the same game as the original, can't use that info to match a game to a netplay lobby or cheat file (and note, the retroachievents project is entirely separate from RetroArch, which is ironic because they have been making progress in I'd ing games in RetroArch because they actually use that info), and retarded some needed changes\caused huge amounts of unnecessary user work in some platform systems to scan games (scummvm only recently got fixed by the maintainer of the core to avoid the scan altogether to create the playlists).

In fact, the only systems where I would not recommend using the manual scanner are MAME (the alternative is to download the mame dat to get the correct names because in this case the filename is not approximately equal to the game name, forbidden characters aside), and scummvm where the best names are got from starting the core, adding the games inside it and going to the global options libretro tab - although I like my own program libretro-scummvm-playlist because it allows me to create multiple scummvm playlists because some engines just have too many games for a single one).
